Add header to the quick edit when bulk editing

louwie17 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This partially addresses #67344, by adding a header to the quick edit panel when we are quick editing multiple pages.

**Follow ups**
- "Mixed" support within fields is done as part of https://github.com/WordPress/gutenberg/pull/67344
- Adding ellipsis dropdown menu with quick actions
     - This will actually be a bit trickier as we have multiple places with similar logic around these actions
     - `PostActions` ( [link](https://github.com/WordPress/gutenberg/blob/3a4cc107ac5c1eeb1c863f0e048cd92ef559bdd2/packages/editor/src/components/post-actions/index.js#L23) ) do we change this to add bulk support?
     - `DataView` bulk actions ( [link](https://github.com/WordPress/gutenberg/blob/81704424da4a17d22e9fcee2a5fe2fc93c8382a1/packages/dataviews/src/components/dataviews-bulk-actions/index.tsx#L120) )

## Why?

In line with the header we show for editing a single page, we should also show one for editing multiple pages.

## How?

I created a `PostEditHeader` component that handles both the header for a single post/page or multiple.
Incase of a single it renders the existing `PostCardPanel`.

## Testing Instructions

1. Enable the "Quick Edit in DataViews" experiment
2. Enable the 2024 or 2025 theme and go to **Appearance > Editor > Pages**
3. Change the view in the top right to **Table** and open the right side bar
4. Select multiple pages, it should show "2 pages" in the header with a `Changes will be applied to all selected pages.` description
5. Toggle back to a single page and notice how it now displays the title of the selected page.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->


https://github.com/user-attachments/assets/3345ef09-8736-49be-947e-ee18c410180b

louwie17 committed:
Add quick edit bulk header

louwie17 commented:
@jameskoster here is the initial header for bulk editing without the ellipsis action menu. Let me know if you have any styling suggestions.

jameskoster commented on packages/edit-site/src/components/post-edit-header/index.js:
```diff
@@ -0,0 +1,84 @@
+					className="edit-site-post-edit-header__title"
+					weight={ 500 }
+					as="h2"
+					lineHeight="20px"
```
Ideally we use [variables](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-menu--with-submenu) or [mixins](https://github.com/WordPress/gutenberg/blob/5efeef9ae740d2aa51fa2493782773083b5363ee/packages/base-styles/_mixins.scss#L18-L70) for text styles rather than hard-coded values. Do you think it's feasible to apply `heading-medium()`?

louwie17 replied:
Sounds good, I updated that as part of: https://github.com/WordPress/gutenberg/pull/67390/commits/f35707b8a2aedb014fa1cb87f25027a27a2ffee0
I did have to update the mixins though, as they were relying on a line height variable that didn't exist, let me know if I updated it correctly: https://github.com/WordPress/gutenberg/pull/67390/files#diff-96351000444fb28f5c80a41909d185066c8202d23cdb19c69fc21f79f76960e8L18-R64

Also I did mimic the above from the `PostCardPanel` component https://github.com/WordPress/gutenberg/blob/f7a4f3f054e99e204b76393293fc4775b9596313/packages/editor/src/components/post-card-panel/index.js#L83 would it make sense to update that header to use the `heading-medium()` as well?

jameskoster replied:
> as they were relying on a line height variable that didn't exist

Oh good catch :D Thanks for fixing that. I guess this is the first time we're using the mixins :)

> would it make sense to update that header to use the heading-medium() as well?

Yes I think they should be the same.

jameskoster commented on packages/edit-site/src/components/post-edit-header/style.scss:
```diff
@@ -0,0 +1,9 @@
+	}
+
+	.edit-site-post-edit-header__title {
+		padding: 2px 0;
```
Is this to align the icon and title? I _think_ we can do that by adjusting the `align` property on the `HStack`, then remove this style.

louwie17 replied:
Yeah, and thanks for calling out the use of `align`. Align `center` allows me to remove the above :)

jameskoster commented:
This is working well :)

I left a couple of small comments.

We might consider using `heading-large()` in `PostEditHeader`, the title looks a little small currently. What do you think?

jameskoster commented:
Random observation, on this branch I'm not seeing the 'Homepage' and 'Posts page' badges next to the relevant page titles. Unsure if it's related to the changes here or not? Seems to be fine on trunk:

<img width="429" alt="Screenshot 2024-11-28 at 16 57 41" src="https://github.com/user-attachments/assets/991a6307-45c8-4e97-b7a3-0249e7659015">

youknowriad commented on packages/edit-site/src/components/post-edit-header/index.js:
```diff
@@ -0,0 +1,84 @@
+			);
+
+			return {
+				icon: unlock( select( editorStore ) ).getPostIcon( postType, {
```
Weird place for such a selector (nothing related to this PR just an observation :P)

youknowriad commented on packages/edit-site/src/components/post-edit-header/index.js:
```diff
@@ -0,0 +1,84 @@
+			const _record = getEditedEntityRecord(
+				'postType',
+				postType,
+				ids[ 0 ]
```
Looks like this only depends on the first id, so we might not need the useMemo for ids at all

louwie17 replied:
Hmm we do still make use of the `ids` further down, both to decide if to show the single header or to show the count of records being edited.
Or do you mean not passing the entire `ids` list into the `useSelect` dependency array?

Btw I am happy to remove `useMemo` use, as `split` is a low effort task. I followed what was done in the `PostEditForm` component ( [line](https://github.com/WordPress/gutenberg/blob/b4297a7249cb34f063a4b8f37fbefc02660787d3/packages/edit-site/src/components/post-edit/index.js#L35) ).
I could also pass in the list of `ids` into the `Header` component.

youknowriad reviewed (commented):
This is nice, and looking good for me. 

We might want to reuse some of the SASS mixins like jay mentioned.

youknowriad commented on packages/edit-site/src/components/post-edit-header/index.js:
```diff
@@ -0,0 +1,84 @@
+/**
```
Nit: I wonder if this deserves its own folder or if it could be just a `header.js` file within the `post-edit`. 

Personally and given the CSS is smallish, I could have kept it in the same folder.

louwie17 replied:
Yeah I was toying with that in my head as well. I did move it to the `post-edit` folder: https://github.com/WordPress/gutenberg/pull/67390/commits/b4297a7249cb34f063a4b8f37fbefc02660787d3

louwie17 committed:
Updating styling of header

louwie17 committed:
Moved header to post-edit folder

oandregal commented:
> how do I actually get these badges to show? I wasn't able to display them on trunk myself. The change of this PR having caused that is low, but still wanted to double check on my end.

Go to "Settings > Reading" and set the "Homepage" and the "Posts page" to a static page:

<img width="870" alt="Screenshot 2024-12-02 at 18 45 35" src="https://github.com/user-attachments/assets/7f5012ab-15cb-454c-bc0b-357736f9da4d">

louwie17 commented:
> > how do I actually get these badges to show? I wasn't able to display them on trunk myself. The change of this PR having caused that is low, but still wanted to double check on my end.
> 
> Go to "Settings > Reading" and set the "Homepage" and the "Posts page" to a static page:

Thanks @oandregal :)

@jameskoster it did still work for me locally on this branch and `trunk`, I also rebased to the latest of `trunk`, so maybe that fixed it 🤷‍♂️ 

![Screenshot 2024-12-03 at 9 44 21 AM](https://github.com/user-attachments/assets/19b29463-e236-426c-8e65-588dea54f518)

youknowriad commented on packages/base-styles/_mixins.scss:
```diff
@@ -659,7 +670,7 @@
+	outline-offset: calc(#{$widthRatio} * ((-1 * var(--wp-admin-border-width-focus)) / var(--wp-block-editor-iframe-zoom-out-scale, 1)));
 }
 
 @mixin selected-block-focus($widthRatio: 1) {
```
Are all these random CSS changes related to this PR?

louwie17 replied:
Looks like a lot of formatting changes got in, when fixing the mixins at the top of this file.

louwie17 replied:
Fixed: https://github.com/WordPress/gutenberg/pull/67390/commits/a0c5fa9d754bd7772d86c1dbecaea52e3fffb71c

youknowriad commented on packages/edit-site/src/components/post-edit/header.js:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	return (
+		<VStack spacing={ 1 } className="edit-site-post-edit-header">
```
Part of me wonders if `PostCardPanel` should support multiple posts (as this seems to recreate a similar UI)

louwie17 replied:
Yeah I could move it there, I initially decided not to given that `PostCardPanel` was initially written for the single header within the site editor.
Also if we do add support, should it be moved to the `editor` package instead?

louwie17 committed:
Make use of mixin for the post title

louwie17 committed:
Pass postId to the pageTypeBadge hook, for use outside editor

youknowriad reviewed (approved):
A couple things I think we should consider doing:

 - Moving the code to the "PostCardPanel" (and support multiple postIds there)
 - Adding the actions dropdown (the actions already support multiple postIds, and I believe actions are used within PostCardPanel already)

That said, this is an improvement 👍

louwie17 committed:
Fix formatting issues

oandregal commented:
This works well, thank you.

There's the detail of the "header" and the "title field" to have the same information, which becomes too much when they have the badge:

<img width="395" alt="Screenshot 2024-12-04 at 18 06 57" src="https://github.com/user-attachments/assets/05841630-0764-4667-ab09-c4948d15ede9">

@jameskoster suggested removing the title field, and that seems sensible. It's not rendered in the DocumentInspector and there's already a "rename action" that enables users to rename it. Though, if we do so, users would no longer have access to rename titles in bulk:

![Screenshot 2024-12-04 at 18 07 08](https://github.com/user-attachments/assets/a00e8e16-d45e-40a8-9ceb-1eb317be949e)

I think this is fine. If we wanted that we could declare the action as a bulk action I presume.

oandregal commented on packages/editor/src/components/post-card-panel/index.js:
```diff
@@ -64,7 +64,7 @@ export default function PostCardPanel( {
 	);
 
-	const pageTypeBadge = usePageTypeBadge();
+	const pageTypeBadge = usePageTypeBadge( parseInt( postId, 10 ) );
```
Why do we need to parse the postId to an integer? Isn't it already one? Or perhaps the previous `getCurrentPostId` returned an integer and the current method (through props somewhere) returns a string? I also see that we don't need to parse it in other places.

louwie17 replied:
Not in the PostEditHeader component where we render this, but it is an integer in the `PostSummary`.
I will remove the parsing here and move that up one level to the header. As we know its a string there for sure.

youknowriad replied:
Just noting that postId is sometimes a real string (templates and template parts). It's true that quick edit is only used for "posts" and "pages" now but `PostCardPanel` can be used for templates and template parts too.

louwie17 replied:
Yeah I moved it up to the `PostEditHeader`: https://github.com/WordPress/gutenberg/pull/67390/commits/5e2d674c986aaf2109ed455f38927d17124641ba

The `usePageTypeBadge` hook works fine whether passing an integer or string, but does rely on integer for the home/posts page match.

oandregal commented on packages/base-styles/_mixins.scss:
```diff
@@ -18,25 +18,25 @@
 	@include _text-heading();
 	font-size: $font-size-x-small;
-	line-height: $line-height-x-small;
+	line-height: $font-line-height-x-small;
```
Ah, nice catch. It looks like these weren't working since https://github.com/WordPress/gutenberg/pull/65418 ?

louwie17 replied:
Yeah correct, although I don't think the Mixin's where being used anywhere within GB yet.

louwie17 committed:
Move parseInt up

louwie17 committed:
Remove title field

louwie17 commented:
Thanks for the additional review @oandregal, I addressed your comment around the `parseInt` and removed `title` from the quick edit list ( [commit](https://github.com/WordPress/gutenberg/pull/67390/commits/58e34e78eee222768bb5fc3d23c886302f380df9) ) 

I will do a follow up PR on these items that @youknowriad had mentioned:
> - Moving the code to the "PostCardPanel" (and support multiple postIds there)
> - Adding the actions dropdown (the actions already support multiple postIds, and I believe actions are used within PostCardPanel already)

oandregal reviewed (approved):
Thanks, this is looking good!
